File: lib/core/utils/access_control.dart
import '../../features/auth/domain/user_role.dart';
import '../../features/auth/domain/user_session.dart';

class AccessControl {
  static bool canCreate(UserRole role) =>
      role == UserRole.admin ||
      role == UserRole.approval ||
      role == UserRole.editor;

  static bool canUpdate(UserRole role) =>
      role == UserRole.admin || role == UserRole.approval;

  static bool canDelete(UserRole role) => role == UserRole.admin;

  static bool canResolveDuplicates(UserRole role) =>
      role == UserRole.admin || role == UserRole.approval;

  static bool canApprove(UserRole role) =>
      role == UserRole.admin || role == UserRole.approval;

  static bool canViewPending(UserRole role) =>
      role == UserRole.admin ||
      role == UserRole.approval ||
      role == UserRole.editor;

  static bool isEditor(UserRole role) => role == UserRole.editor;

  static bool isApprovalRole(UserRole role) => role == UserRole.approval;

  static bool canViewAllGenders(UserRole role) => role == UserRole.admin;

  static String? genderFilter(UserSession? session) {
    if (session == null) return null;
    if (canViewAllGenders(session.role)) return null;
    return session.gender;
  }

  static bool canSetGender(UserSession session) =>
      session.role == UserRole.admin;

  static String effectiveGender(UserSession session, String selectedGender) {
    if (session.role == UserRole.admin) {
      return UserSession.normalizeGender(selectedGender);
    }
    return UserSession.normalizeGender(session.gender);
  }
}

File: app/src/main/java/com/example/salik_management_system/core/utils/AccessControl.kt
package com.example.salik_management_system.core.utils

import com.example.salik_management_system.auth.domain.UserRole
import com.example.salik_management_system.auth.domain.UserSession

object AccessControl {
    fun canCreate(role: UserRole): Boolean =
        role == UserRole.Admin || role == UserRole.Approval || role == UserRole.Editor

    fun canUpdate(role: UserRole): Boolean =
        role == UserRole.Admin || role == UserRole.Approval

    fun canDelete(role: UserRole): Boolean = role == UserRole.Admin

    fun canResolveDuplicates(role: UserRole): Boolean =
        role == UserRole.Admin || role == UserRole.Approval

    fun canApprove(role: UserRole): Boolean =
        role == UserRole.Admin || role == UserRole.Approval

    fun canViewPending(role: UserRole): Boolean =
        role == UserRole.Admin || role == UserRole.Approval || role == UserRole.Editor

    fun isEditor(role: UserRole): Boolean = role == UserRole.Editor

    fun isApprovalRole(role: UserRole): Boolean = role == UserRole.Approval

    fun canViewAllGenders(role: UserRole): Boolean = role == UserRole.Admin

    fun genderFilter(session: UserSession?): String? {
        if (session == null) return null
        if (canViewAllGenders(session.role)) return null
        return session.gender
    }

    fun canSetGender(session: UserSession): Boolean = session.role == UserRole.Admin

    fun effectiveGender(session: UserSession, selectedGender: String): String {
        return if (session.role == UserRole.Admin) {
            UserSession.normalizeGender(selectedGender)
        } else {
            UserSession.normalizeGender(session.gender)
        }
    }
}
